src/Tracer.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Exercism\PhpTestRunner;
+
+use Exercism\PhpTestRunner\Result;
+use PHPUnit\Event\Code\TestMethod;
+use PHPUnit\Event\Event;
+use PHPUnit\Event\Test\BeforeFirstTestMethodErrored;
+use PHPUnit\Event\Test\Errored;
+use PHPUnit\Event\Test\Failed;
+use PHPUnit\Event\Test\Passed;
+use PHPUnit\Event\Test\PrintedUnexpectedOutput;
+use PHPUnit\Event\TestRunner\Finished;
+use PHPUnit\Event\Tracer\Tracer as TracerInterface;
+use ReflectionClass;
+
+final class Tracer implements TracerInterface
+{
+    private array $result = [
+        'version' => 3,
+        'status' => 'pass',
+        'tests' => [],
+    ];
+
+    public function __construct(
+        private readonly string $outFileName,
+        private readonly string $exerciseDir,
+    ) {
+    }
+
+    public function trace(Event $event): void
+    {
+        match (\get_class($event)) {
+            Passed::class => $this->addTestPassed($event),
+            Failed::class => $this->addTestFailed($event),
+            Errored::class => $this->addTestErrored($event),
+            BeforeFirstTestMethodErrored::class => $this->addBeforeFirstTestMethodErrored($event),
+            PrintedUnexpectedOutput::class => $this->addTestOutput($event),
+            // default => $this->addUnhandledEvent($event),
+            default => true,
+        };
+
+
+        if ($event instanceof Finished) {
+            $this->saveResults();
+        }
+    }
+
+    private function addUnhandledEvent(Event $event): void
```

Comment:
Do we want to keep this?

Reply:
Yes, there is very many new events to handle for the list of missing features (PHP warnings, deprecations etc.). This helped me finding the ones coming in and you shouldn't have to re-write it for future development.

Reply:
Maybe we can have this behind a `--debug` flag then? or an env variable. So we don't have commented code?

Reply:
There's 2 variations for this. I'll add constants to toggle these...